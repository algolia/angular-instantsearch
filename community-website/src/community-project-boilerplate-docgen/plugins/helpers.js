const highlight = require('./syntaxHighlighting.js');
const md = require('./mdRenderer.js');

// this plugin provides ATM one helper to easily compute the publicPath of assets
module.exports = function helpers(filenames, metalsmith, cb) {
  // eslint-disable-next-line no-param-reassign
  metalsmith.metadata().h = {
    markdown(src) {
      return md.render(src);
    },
    highlight(src, opts) {
      const lang = opts && opts.lang;
      return highlight(src, lang);
    },
    maybeActive(navPath, singlePathOrArrayOfPaths) {
      const pathsToTest = [].concat(singlePathOrArrayOfPaths);
      return pathsToTest.some(pathToTest => navPath.indexOf(pathToTest) === 0)
        ? 'active'
        : '';
    },
  };

  cb();
};
