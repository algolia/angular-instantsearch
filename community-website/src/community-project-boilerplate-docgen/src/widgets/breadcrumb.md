---
title: Widgets - Breadcrumb
layout: widget.pug
---

# Breadcrumb

## Description

A breadcrumb is a secondary navigation scheme that allows the user to see where the current page is in relation to the website or web applicationâ€™s hierarchy. In terms of usability, using a breadcrumb reduces the number of actions a visitor needs to take in order to get to a higher-level page.

## Options

| Attribute            | Type       | Description
| -                    | -          | -
| `attributes`         | `string[]` | Array of attributes to use to craft the Breadcrumb
| `rootPath?`          | `string`   | Prefix path to use if the first level is not the root level
| `autoHideContainer?` | `boolean`  | Hides the breadcrumb if there's no item to display

## Code example

```html
<ng-ais-breadcrumb
  [attributes]="[
    'hierarchicalCategories.lvl0',
    'hierarchicalCategories.lvl1',
    'hierarchicalCategories.lvl2'
  ]"
>
</ng-ais-breadcrumb>
```
