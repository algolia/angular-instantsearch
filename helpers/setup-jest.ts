import "jest-preset-angular";

// stupid workaround for TS
import * as helper from "algoliasearch-helper";
require("algoliasearch-helper").default = helper;

const mock = () => {
  let storage = {};
  return {
    getItem(key) {
      return storage[key] || null;
    },
    setItem(key, value) {
      storage[key] = value;
    },
    removeItem(key) {
      delete storage[key];
    },
    clear() {
      storage = {};
    }
  };
};

Object.defineProperty(window, "localStorage", { value: mock() });
Object.defineProperty(window, "sessionStorage", { value: mock() });
Object.defineProperty(window, "getComputedStyle", {
  value: () => ["-webkit-appearance"]
});
