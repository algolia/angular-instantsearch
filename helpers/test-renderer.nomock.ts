import { Component, ViewChild } from '@angular/core';
import { TestBed } from '@angular/core/testing';

import { NgAisInstantSearchModule } from '../src/instantsearch/instantsearch.module';

// same as test-renderer, but without mocking InstantSearch
// In the future, we want to migrate all tests to this

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations,
  methods = {},
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
  methods?: {};
}) {
  return function(state?: {}, isFirstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined,
        methods,
      },
      isFirstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state,
    methods,
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
    methods?: {};
  },
  isFirstRender = false
) {
  @Component({
    template: `
      <ais-instantsearch [config]="config">
        ${template}
      </ais-instantsearch>
    `,
  })
  class TestContainer {
    config = {
      searchClient: {
        search(requests) {
          return Promise.resolve({
            results: requests.map(() => ({ hits: [] })),
          });
        },
      },
      indexName: 'instant_search',
    };
    @ViewChild(TestedWidget) testedWidget;
    constructor() {
      Object.keys(methods).forEach(methodName => {
        this[methodName] = methods[methodName];
      });
    }
  }

  TestBed.configureCompiler({
    preserveWhitespaces: false,
  } as any).configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || []),
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])],
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, isFirstRender);
    fixture.detectChanges();
  }

  return fixture;
}
