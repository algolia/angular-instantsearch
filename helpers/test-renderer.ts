import { Component, ViewChild } from "@angular/core";
import { TestBed } from "@angular/core/testing";

import { NgAisInstantSearchModule } from "../src/instantsearch/instantsearch.module";

// mock component, dont create a real instantearch instance
jest.mock("../src/base-widget");
jest.mock("../src/instantsearch/instantsearch");

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
}) {
  return function(state?: {}, firstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined
      },
      firstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
  },
  firstRender = false
) {
  @Component({
    template: `
      <ng-ais-instantsearch>
        ${template}
      </ng-ais-instantsearch>
    `
  })
  class TestContainer {
    @ViewChild(TestedWidget) testedWidget;
  }

  TestBed.configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || [])
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])]
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, firstRender);
    fixture.detectChanges();
  }

  return fixture;
}
