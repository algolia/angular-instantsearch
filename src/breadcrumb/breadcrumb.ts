import { Component, Input, Inject, forwardRef } from '@angular/core';
import { connectBreadcrumb } from 'instantsearch.js/es/connectors';
import { BaseWidget } from '../base-widget';
import { NgAisInstantSearch } from '../instantsearch/instantsearch';
import { noop } from '../utils';

export type BreadcrumbState = {
  createURL: Function;
  items: BreadcrumbItem[];
  refine: Function;
};

export type BreadcrumbItem = {
  name: string;
  value: string;
};

@Component({
  selector: 'ais-breadcrumb',
  template: `
    <div
      [class]="cx()"
      *ngIf="!isHidden"
    >
      <ul [class]="cx('list')">
        <li
          *ngFor="let item of items"
          [class]="cx('item', item.isLast ? 'selected' : undefined)"
          (click)="handleClick($event, item)"
        >
          <span
            *ngIf="item.separator"
            [class]="cx('separator')"
            aria-hidden="true"
          >
            >
          </span>
          <a
            [class]="cx('link')"
            href="{{state.createURL(item.value)}}"
            *ngIf="!item.isLast"
            (click)="handleClick($event, item)"
          >
            {{item.label}}
          </a>

          <span *ngIf="item.isLast">
            {{item.label}}
          </span>
        </li>
      </ul>
    </div>
  `,
})
export class NgAisBreadcrumb extends BaseWidget {
  // instance options
  @Input() public attributes: string[];
  @Input() public rootPath?: string;
  // TODO: add separator?
  // TODO: add transformItems?

  get isHidden() {
    return this.state.items.length === 0 && this.autoHideContainer;
  }

  get items() {
    return this.state.items.map((item, idx) => ({
      ...item,
      separator: idx !== 0,
      isLast: idx === this.state.items.length - 1,
      // FIXME: get rid of this. We can use `last` local variable
      // https://angular.io/api/common/NgForOf#local-variables
    }));
  }

  public state: BreadcrumbState = {
    createURL: noop,
    items: [],
    refine: noop,
  };

  constructor(
    @Inject(forwardRef(() => NgAisInstantSearch))
    public instantSearchParent: any
  ) {
    super('Breadcrumb');
  }

  public ngOnInit() {
    this.createWidget(connectBreadcrumb, {
      attributes: this.attributes,
      rootPath: this.rootPath,
    });

    super.ngOnInit();
  }

  public handleClick(event: MouseEvent, item: BreadcrumbItem) {
    event.preventDefault();
    event.stopPropagation();

    if (item.value) {
      this.state.refine(item.value);
    }
  }
}
