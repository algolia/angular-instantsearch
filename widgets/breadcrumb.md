## `<ng-ais-breadcrumb />`

A breadcrumb is a secondary navigation scheme that allows the user to see where the current page is in relation to the website or web applicationâ€™s hierarchy. In terms of usability, using a breadcrumb reduces the number of actions a visitor needs to take in order to get to a higher-level page.

| Attribute          | Type       | Description
| -                  | -          | -
| `attributes`       | `string[]` | Array of attributes to use to craft the Breadcrumb
| `rootPath?`        | `string`   | Prefix path to use if the first level is not the root level
| `header?`          | `string`   | Displays text before widget
| `footer?`          | `string`   | Displays text after widget