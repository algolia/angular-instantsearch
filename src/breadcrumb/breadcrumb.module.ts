import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";

import { NgAisBreadcrumb } from "./breadcrumb";

@NgModule({
  declarations: [NgAisBreadcrumb],
  entryComponents: [NgAisBreadcrumb],
  exports: [NgAisBreadcrumb],
  imports: [CommonModule]
})
export class NgAisBreadcrumbModule {}
