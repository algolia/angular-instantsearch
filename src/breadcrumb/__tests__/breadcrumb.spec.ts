import { createRenderer } from '../../../helpers/test-renderer';
import { NgAisBreadcrumb } from '../breadcrumb';

const defaultState = {
  createURL: jest.fn(),
  items: [{ name: 'foo', value: 'foo' }, { name: 'bar', value: 'bar' }],
  refine: jest.fn(),
};

const render = createRenderer({
  defaultState,
  template: '<ais-breadcrumb></ais-breadcrumb>',
  TestedWidget: NgAisBreadcrumb,
});

describe('Breadcrumb', () => {
  it('renders markup without state', () => {
    const fixture = render();
    expect(fixture).toMatchSnapshot();
  });

  it('renders markup with two items', () => {
    const fixture = render({});
    expect(fixture).toMatchSnapshot();
  });

  it('should refine when clicking on first item', () => {
    const refine = jest.fn();
    const fixture = render({ refine });

    const firstLink = fixture.debugElement.nativeElement.querySelector(
      'a:first-child'
    );

    firstLink.click();

    expect(refine).toHaveBeenCalled();
    expect(refine).toHaveBeenCalledWith('foo');
  });

  it('should be hidden with `autoHideContainer`', () => {
    const fixture = render({ items: [] });
    fixture.componentInstance.testedWidget.autoHideContainer = true;
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });
});
