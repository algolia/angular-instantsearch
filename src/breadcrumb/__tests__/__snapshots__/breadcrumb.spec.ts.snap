// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb renders markup with two items 1`] = `
<ng-component
  testedWidget={[Function NgAisBreadcrumb]}
>
  
        <ng-ais-instantsearch>
    
          
    <ng-ais-breadcrumb>
      
      
      
      <div
        class="ais-Breadcrumb"
      >
        
        
        <ul
          class="ais-Breadcrumb-list"
        >
          
          
          
          <li
            class="ais-Breadcrumb-item"
          >
            
            
            
            
            
            
            <a
              class="ais-Breadcrumb-link"
              href=""
            >
              
              foo
            
            </a>
            
  
            
            
            
          
          </li>
          <li
            class="ais-Breadcrumb-item--selected"
          >
            
            
            
            <span
              aria-hidden="true"
              class="ais-Breadcrumb-separator"
            >
              
              &gt;
            
            </span>
            
            
            
            
  
            
            
            <span>
              
              bar
            
            </span>
            
          
          </li>
          
        
        </ul>
        
      
      </div>
      
    
    </ng-ais-breadcrumb>
    
        
  </ng-ais-instantsearch>
      
</ng-component>
`;

exports[`Breadcrumb renders markup without state 1`] = `
<ng-component
  testedWidget={[Function NgAisBreadcrumb]}
>
  
        <ng-ais-instantsearch>
    
          
    <ng-ais-breadcrumb>
      
      
      
      <div
        class="ais-Breadcrumb"
      >
        
        
        <ul
          class="ais-Breadcrumb-list"
        >
          
          
          
          
        
        </ul>
        
      
      </div>
      
    
    </ng-ais-breadcrumb>
    
        
  </ng-ais-instantsearch>
      
</ng-component>
`;

exports[`Breadcrumb should be hidden with \`autoHideContainer\` 1`] = `
<ng-component
  testedWidget={[Function NgAisBreadcrumb]}
>
  
        <ng-ais-instantsearch>
    
          
    <ng-ais-breadcrumb>
      
      
      
      
    
    </ng-ais-breadcrumb>
    
        
  </ng-ais-instantsearch>
      
</ng-component>
`;
